man/systemd-udevd.service.xml

```diff
@@ -160,6 +160,16 @@
           terminated due to kernel drivers taking too long to initialize.</para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>udev.blockdev_read_only</varname></term>
+        <listitem>
+          <para>If specified, mark all physical block devices read-only as they appear. Synthetic block
+          devices (such as loopback block devices or device mapper devices) are left as they are. This is
+          useful to guarantee that the contents of physical block devices remains unmodified during runtime,
+          for example to implement fully stateless systems, for testing or for recovery situations where
+          corrupted file systems shall not be corrupted further through accidental modifications.</para>
```

Comment:
It would be nice to give a hint how to mark specific devices as rw again.

EDIT: Actually, is it even possible? After marking the device rw again, `mount -o remount,rw` fails with `mount: /: cannot remount /dev/vda3 read-write, is write-protected.`. I used a small C program to call `ioctl(..., BLKROSET, [0])` on both /dev/vda and /dev/vda3. vda3 has btrfs.

EDIT: Oh, it's a btrfs "feature".

Reply:
might be something to inform the btrfs folks about that would be good to get fixed

Reply:
https://bugzilla.kernel.org/show_bug.cgi?id=208075